### PR TITLE
Improved support for BINARY(n) and VARBINARY(n) columns

### PR DESCRIPTION
**Pull request by dmromanov**
Improved support for BINARY(n) and VARBINARY(n) columns.

### CODE REVIEW

File: src/Database/Schema/SqliteSchema.php [hunk 1]
```diff
@@ -340,6 +339,14 @@ public function columnSql(TableSchema $schema, $name)
             }
         }
 
+        if ($data['type'] === TableSchema::TYPE_BINARY) {
+            if (isset($data['length'])) {
+                $out .= ' VARBINARY(' . (int)$data['length'] . ')';
+            } else {
+                $out .= ' BLOB';
+            }
```
Comment (markstory) [inline src/Database/Schema/SqliteSchema.php]: I don't know if these types are actually different in sqlite. They might be cosmetically different  but behave the same.

File: src/Database/Schema/SqlserverSchema.php [hunk 2]
```diff
@@ -371,12 +371,26 @@ public function columnSql(TableSchema $schema, $name)
...
-                $out .= sprintf('(%s)', TableSchema::LENGTH_TINY);
+                $out .= ' VARBINARY';
+
+                if (empty($data['length']) || in_array(
+                    $data['length'],
+                    [TableSchema::LENGTH_MEDIUM, TableSchema::LENGTH_LONG],
+                    true
+                )) {
+                    $out .= '(MAX)';
+                } else {
```
Comment (markstory) [inline src/Database/Schema/SqlserverSchema.php]: This first of condition could be folded into the if ladder inside of it avoiding a layer of nesting.

File: tests/TestCase/Database/Schema/PostgresSchemaTest.php [hunk 3]
```diff
@@ -203,6 +203,15 @@ public static function convertColumnProvider()
...
                 ['type' => 'binary', 'length' => null]
             ],
+            // Binary
+            [
+                ['type' => 'BIT VARYING(3)'],
+                ['type' => 'binary', 'length' => 3]
+            ],
+            [
+                ['type' => 'BIT(1)'],
+                ['type' => 'binary', 'length' => 1]
```
Comment (markstory) [inline tests/TestCase/Database/Schema/PostgresSchemaTest.php]: Bit fields in postgres are odd as they are represented with strings. I am not sure mapping them to binary will give folks the results they are expecting.

Comment (dmromanov) [reply tests/TestCase/Database/Schema/PostgresSchemaTest.php]: Sounds reasonable

File: src/Database/Schema/SqlserverSchema.php (hunk 2 above)
Comment (dmromanov) [reply src/Database/Schema/SqlserverSchema.php]: Done.

File: src/Database/Schema/MysqlSchema.php [hunk 4]
```diff
@@ -359,7 +359,17 @@ public function columnSql(TableSchema $schema, $name)
                     break;
                 case TableSchema::TYPE_BINARY:
                     $isKnownLength = in_array($data['length'], TableSchema::$columnLengths);
-                    if (empty($data['length']) || !$isKnownLength) {
+
+                    if (!empty($data['length']) && !$isKnownLength) {
+                        if ($data['length'] > 2) {
+                            $out .= ' VARBINARY(' . strtoupper($data['length']) . ')';
+                        } else {
+                            $out .= ' BINARY(' . strtoupper($data['length']) . ')';
```
Comment (markstory) [inline src/Database/Schema/MysqlSchema.php]: Does `strtoupper()` on a number do anything?

Comment (dmromanov) [reply src/Database/Schema/MysqlSchema.php]: Corrected :)

**Pull request merged**